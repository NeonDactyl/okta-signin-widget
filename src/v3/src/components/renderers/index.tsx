/*
 * Copyright (c) 2022-present, Okta, Inc. and/or its affiliates. All rights reserved.
 * The Okta software accompanied by this notice is provided pursuant to the Apache License, Version 2.0 (the "License.")
 *
 * You may obtain a copy of the License at http://www.apache.org/licenses/LICENSE-2.0.
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
 * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 *
 * See the License for the specific language governing permissions and limitations under the License.
 */

import {
  and,
  isBooleanControl,
  isEnumControl,
  isStringControl,
  optionIs,
  rankWith,
  scopeEndIs,
  scopeEndsWith,
  uiTypeIs,
} from '@jsonforms/core';
import {
  withJsonFormsCellProps,
  withJsonFormsContext,
  withJsonFormsControlProps,
} from '@jsonforms/react';
import { vanillaRenderers } from '@jsonforms/vanilla-renderers';

import { withFormValidationState, withFormValidationStateAndContext } from '../hocs';
import
AuthenticatorListControl,
{ tester as authenticatorListControlTester } from './authenticatorListControl/AuthenticatorListControl';
import ButtonControl, { buttonControlTester } from './buttonControl/ButtonControl';
import WebAuthNSubmitControl from './buttonControl/WebAuthNSubmitControl';
import SuccessCallbackControl from './callbackControl/SuccessCallbackControl';
import CheckboxControl from './checkboxControl/CheckboxControl';
import FastPassButtonControl, { tester as fastPassButtonControlTester } from './fastPassButtonControl/FastPassButtonControl';
import ImageWithTextControl from './imageWithTextControl/ImageWithTextControl';
import InfoBoxControl from './infoBoxControl/InfoBoxControl';
import InformationalTextControl from './informationalTextControl/informationalTextControl';
import InputPasswordControl from './inputPasswordControl/InputPasswordControl';
import InputTextControl from './inputTextControl/InputTextControl';
import InputTextHiddenControl from './inputTextHiddenControl/inputTextHiddenControl';
import InputVerificationCodeControl from './inputVerificationCodeControl/InputVerificationCodeControl';
import LinkControl from './linkControl/LinkControl';
import ListControl from './listControl/ListControl';
import PasswordRequirementsControl from './passwordRequirementsControl/PasswordRequirementsControl';
import PhoneAuthenticatorControl from './phoneAuthControl/PhoneAuthenticatorControl';
import QrCodeControl from './qrCodeControl/QrCodeControl';
import RadioControl from './radioControl/RadioControl';
import RedirectControl from './redirectControl/RedirectControl';
import ReminderPromptControl from './reminderPromptControl/ReminderPromptControl';
import SelectControl from './selectControl/SelectControl';
import SpinnerControl from './spinner/Spinner';
import StepperRenderer from './stepperRenderer/StepperRenderer';
import TitleControl from './titleControl/titleControl';

export const renderers = [
  ...vanillaRenderers,
  {
    tester: rankWith(3, scopeEndsWith('phoneNumber')),
    renderer: withJsonFormsControlProps(
      withFormValidationStateAndContext(withJsonFormsContext(PhoneAuthenticatorControl)),
    ),
  },
  {
    tester: rankWith(3, isStringControl),
    renderer: withJsonFormsControlProps(withFormValidationState(InputTextControl)),
  },
  {
    tester: rankWith(3, isBooleanControl),
    renderer: withJsonFormsControlProps(withFormValidationState(CheckboxControl)),
  },
  {
    tester: rankWith(4, uiTypeIs('Title')),
    renderer: withJsonFormsCellProps(TitleControl),
  },
  {
    tester: rankWith(4, uiTypeIs('Description')),
    renderer: withJsonFormsCellProps(InformationalTextControl),
  },
  {
    tester: rankWith(6, and(isEnumControl, optionIs('format', 'radio'))),
    renderer: withJsonFormsControlProps(RadioControl),
  },
  {
    tester: rankWith(6, and(isEnumControl, optionIs('format', 'dropdown'))),
    renderer: withJsonFormsControlProps(withFormValidationState(SelectControl)),
  },
  {
    tester: rankWith(7, authenticatorListControlTester),
    renderer: withJsonFormsControlProps(withJsonFormsContext(AuthenticatorListControl)),
  },
  {
    tester: rankWith(9, scopeEndIs('verificationCode')),
    renderer: withJsonFormsControlProps(withFormValidationState(InputVerificationCodeControl)),
  },
  {
    tester: rankWith(9, uiTypeIs('Spinner')),
    renderer: withJsonFormsCellProps(SpinnerControl),
  },
  {
    tester: rankWith(9, uiTypeIs('Button')),
    // TODO: create new renderer function per OKTA-473161
    renderer: withJsonFormsCellProps(WebAuthNSubmitControl),
  },
  {
    tester: rankWith(9, uiTypeIs('InfoBox')),
    renderer: withJsonFormsCellProps(InfoBoxControl),
  },
  {
    tester: rankWith(9, uiTypeIs('Link')),
    renderer: withJsonFormsCellProps(LinkControl),
  },
  {
    tester: rankWith(9, uiTypeIs('List')),
    renderer: withJsonFormsCellProps(ListControl),
  },
  {
    tester: rankWith(9, uiTypeIs('QRCode')),
    renderer: withJsonFormsCellProps(QrCodeControl),
  },
  {
    tester: rankWith(10, optionIs('secret', true)),
    renderer: withJsonFormsControlProps(withFormValidationState(InputPasswordControl)),
  },
  {
    tester: rankWith(10, optionIs('format', 'ImageWithText')),
    renderer: withJsonFormsCellProps(ImageWithTextControl),
  },
  {
    tester: buttonControlTester,
    renderer: withJsonFormsControlProps(ButtonControl),
  },
  {
    tester: fastPassButtonControlTester,
    renderer: withJsonFormsControlProps(FastPassButtonControl),
  },
  {
    tester: rankWith(11, uiTypeIs('Stepper')),
    renderer: StepperRenderer,
  },
  {
    tester: rankWith(11, uiTypeIs('Reminder')),
    renderer: withJsonFormsCellProps(ReminderPromptControl),
  },
  {
    tester: rankWith(11, uiTypeIs('PasswordRequirements')),
    renderer: withJsonFormsCellProps(withJsonFormsContext(PasswordRequirementsControl)),
  },
  {
    tester: rankWith(11, uiTypeIs('SuccessCallback')),
    renderer: withJsonFormsCellProps(SuccessCallbackControl),
  },
  {
    tester: rankWith(11, uiTypeIs('Redirect')),
    renderer: withJsonFormsCellProps(RedirectControl),
  },
  {
    tester: rankWith(12, optionIs('type', 'hidden')),
    renderer: withJsonFormsControlProps(InputTextHiddenControl),
  },
];
