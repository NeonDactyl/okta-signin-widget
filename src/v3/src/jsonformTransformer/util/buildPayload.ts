/*
 * Copyright (c) 2022-present, Okta, Inc. and/or its affiliates. All rights reserved.
 * The Okta software accompanied by this notice is provided pursuant to the Apache License, Version 2.0 (the "License.")
 *
 * You may obtain a copy of the License at http://www.apache.org/licenses/LICENSE-2.0.
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
 * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 *
 * See the License for the specific language governing permissions and limitations under the License.
 */

import { IdxActionParams } from '@okta/okta-auth-js';
import { merge } from 'lodash';

import { FormBag } from '../../types';
import { toNestedObject } from './toNestedObject';

export const buildPayload = (
  formbag: FormBag,
  params?: IdxActionParams,
): IdxActionParams => {
  const unmutableData = Object.entries(formbag.schema.properties || {})
    // TODO: figure out best way to pass idx meta to json form schema
    // @ts-ignore
    .filter(([_key, value]) => value.mutable === false)
    .reduce((acc, [key, value]) => {
      const obj = toNestedObject(key, value.const);
      return merge(acc, obj);
    }, {});
  return merge(unmutableData, params);
};
