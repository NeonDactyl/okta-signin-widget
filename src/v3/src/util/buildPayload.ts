/*
 * Copyright (c) 2022-present, Okta, Inc. and/or its affiliates. All rights reserved.
 * The Okta software accompanied by this notice is provided pursuant to the Apache License, Version 2.0 (the "License.")
 *
 * You may obtain a copy of the License at http://www.apache.org/licenses/LICENSE-2.0.
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
 * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 *
 * See the License for the specific language governing permissions and limitations under the License.
 */

import { JsonSchema7 } from '@jsonforms/core';
import { IdxActionParams } from '@okta/okta-auth-js';
import merge from 'lodash/merge';
import set from 'lodash/set';

import { FormBag, JsonObject, Undefinable } from '../types';

const parseProperties = (properties: JsonSchema7['properties'], path?: string[]): JsonObject => {
  const immutableData = Object.entries(properties || {})
    .reduce((acc, [key, val]) => {
      const objectPathArr: string[] = [...(path ?? []).filter(Boolean), key];
      if (val.properties) {
        return parseProperties(val.properties, objectPathArr);
      }

      if (val.readOnly) {
        set(acc, objectPathArr, val.const);
      }
      return acc;
    }, {});
  return immutableData;
};

export const buildPayload = (
  formBag?: FormBag,
  params?: IdxActionParams,
): Undefinable<IdxActionParams> => {
  if (!formBag) {
    return params;
  }

  const immutableData = parseProperties(formBag.schema.properties || {});

  return merge(immutableData, params);
};
