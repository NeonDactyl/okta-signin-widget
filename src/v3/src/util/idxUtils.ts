/*
 * Copyright (c) 2022-present, Okta, Inc. and/or its affiliates. All rights reserved.
 * The Okta software accompanied by this notice is provided pursuant to the Apache License, Version 2.0 (the "License.")
 *
 * You may obtain a copy of the License at http://www.apache.org/licenses/LICENSE-2.0.
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
 * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 *
 * See the License for the specific language governing permissions and limitations under the License.
 */

import { IdxMessage, IdxRemediation, IdxTransaction } from '@okta/okta-auth-js';
import {
  AuthCoinProps,
  Nullable,
  RequiredKeys,
  Undefinable,
  UserInfo,
  WidgetProps,
} from 'src/types';
import { AppInfo } from 'src/types/appInfo';

import {
  AUTHENTICATOR_KEY,
  EMAIL_AUTHENTICATOR_TERMINAL_KEYS,
  IDX_STEP,
} from '../constants';

export const getUserInfo = (transaction: IdxTransaction): UserInfo => {
  const { context: { user } } = transaction;

  if (!user) {
    return {};
  }
  return user.value as UserInfo;
};

export const getAppInfo = (transaction: IdxTransaction): AppInfo => {
  const { context: { app } } = transaction;

  if (!app) {
    return {};
  }
  return app.value as AppInfo;
};

export const containsMessageKey = (
  key: string,
  messages?: IdxMessage[],
): boolean => (messages?.some((message) => message.i18n?.key === key) ?? false);

export const containsMessageKeyPrefix = (
  prefix: string,
  messages?: IdxMessage[],
): boolean => (messages?.some((message) => message.i18n?.key?.startsWith(prefix)) ?? false);

export const containsOneOfMessageKeys = (
  keys: string[],
  messages?: IdxMessage[],
): boolean => keys.some((key) => containsMessageKey(key, messages));

export const buildAuthCoinProps = (
  transaction?: Nullable<IdxTransaction>,
): Undefinable<AuthCoinProps> => {
  if (!transaction) {
    return undefined;
  }

  const { nextStep, messages } = transaction;
  if (containsOneOfMessageKeys(EMAIL_AUTHENTICATOR_TERMINAL_KEYS, messages)
    || nextStep?.name === IDX_STEP.EMAIL_CHALLENGE_CONSENT) {
    return { authenticatorKey: AUTHENTICATOR_KEY.EMAIL };
  }

  if (!nextStep?.authenticator?.key) {
    return undefined;
  }

  return { authenticatorKey: nextStep.authenticator.key };
};

export const hasMinAuthenticatorOptions = (
  transaction: IdxTransaction,
  stepName: string,
  min: number,
): boolean => {
  const excludedPages = [
    IDX_STEP.SELECT_AUTHENTICATOR_ENROLL,
    IDX_STEP.SELECT_AUTHENTICATOR_AUTHENTICATE,
  ];
  if (excludedPages.includes(transaction.nextStep?.name ?? '')) {
    return false;
  }

  const step: Undefinable<IdxRemediation> = transaction.neededToProceed.find(
    ({ name }) => name === stepName,
  );
  if (!step) {
    return false;
  }

  return (step.value?.find(({ name }) => name === 'authenticator')?.options?.length ?? 0) > min;
};

export const isAuthClientSet = (
  props: WidgetProps,
): props is RequiredKeys<WidgetProps, 'authClient'> => !!props.authClient;
