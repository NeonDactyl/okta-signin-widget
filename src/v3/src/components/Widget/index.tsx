/*
 * Copyright (c) 2022-present, Okta, Inc. and/or its affiliates. All rights reserved.
 * The Okta software accompanied by this notice is provided pursuant to the Apache License, Version 2.0 (the "License.")
 *
 * You may obtain a copy of the License at http://www.apache.org/licenses/LICENSE-2.0.
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
 * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 *
 * See the License for the specific language governing permissions and limitations under the License.
 */

import { JsonFormsCore } from '@jsonforms/core';
import { JsonForms } from '@jsonforms/react';
import {
  vanillaCells,
} from '@jsonforms/vanilla-renderers';
import { ThemeProvider } from '@okta/odyssey-react-theme';
import {
  APIError,
  IdxMessage,
  IdxStatus,
  IdxTransaction,
} from '@okta/okta-auth-js';
import Ajv, { ErrorObject } from 'ajv';
import AjvErrors from 'ajv-errors';
import debounce from 'lodash/debounce';
import { FunctionComponent, h } from 'preact';
import {
  useCallback,
  useEffect,
  useMemo,
  useState,
} from 'preact/hooks';
import {
  FormBag,
  IdxTransactionWithNextStep,
  Nullish,
  Undefinable,
} from 'src/types/jsonforms';
import { WidgetProps } from 'src/types/widget';

import { WidgetContextProvider } from '../../contexts';
import { usePolling } from '../../hooks';
import transformTransaction from '../../transformer/authJs';
import { transformTerminalTransaction, transformUnhandledErrors } from '../../transformer/terminal';
import { buildAuthCoinProps, buildPayload, isAuthClientSet } from '../../util';
import { mapThemeFromBrand } from '../../util/theme';
import AuthContainer from '../AuthContainer/AuthContainer';
import AuthContent from '../AuthContent/AuthContent';
import AuthHeader from '../AuthHeader/AuthHeader';
import InfoSection from '../InfoSection/InfoSection';
import { renderers } from '../renderers';
import Spinner from '../Spinner';

const ajv = new Ajv({
  schemaId: 'auto',
  // errorDataPath is required to be 'property' for JsonForms to pass error messages
  // to individual component renderers correctly
  // However, this breaks custom error messages via ajv-errors
  // For now, we will keep it this way and solve custom errors later
  // TODO: OKTA-462462
  errorDataPath: 'property',
  allErrors: true,
  jsonPointers: true,
  verbose: true,
  $data: true,
});
AjvErrors(ajv, {});

export const Widget: FunctionComponent<WidgetProps> = (widgetProps) => {
  if (!isAuthClientSet(widgetProps)) {
    throw new Error('authClient is required');
  }

  const {
    authClient,
    brandColors,
    brandName,
    logo,
    logoText,
    onSuccess,
    stateToken,
  } = widgetProps;

  const TERMINAL_STATUSES = [IdxStatus.TERMINAL, IdxStatus.SUCCESS];

  const [data, setData] = useState({});
  // TODO OKTA-462462 upgrade to JSONForms@3.x
  // TODO OKTA-432181 custom client-side error messages
  // eslint-disable-next-line @typescript-eslint/no-unused-vars
  const [formErrors, setFormErrors] = useState<ErrorObject[]>([]);
  const [messages, setMessages] = useState<IdxMessage[]>([]);
  const [idxTransaction, setIdxTransaction] = useState<Nullish<IdxTransaction>>();
  const [authApiError, setAuthApiError] = useState<Undefinable<APIError>>();
  const pollingTransaction = usePolling(idxTransaction);

  const bootstrap = useCallback(async () => {
    const transaction = await authClient.idx.start({
      stateHandle: stateToken,
    }).catch((error) => {
      // TODO: handle error based on types
      // AuthApiError is one of the potential error that can be thrown here
      // We will want to expose development stage errors from auth-js and file jiras against it
      setAuthApiError(error);
      console.error(error);
      return null;
    });
    setIdxTransaction(transaction);
  }, [authClient, stateToken, setIdxTransaction]);

  const resume = useCallback(async () => {
    const transaction = await authClient.idx.proceed({
      stateHandle: idxTransaction?.context.stateHandle,
    })
      .catch((error) => {
        setAuthApiError(error);
        console.error(error);
        return null;
      });
    setIdxTransaction(transaction);
  }, [authClient, setIdxTransaction]);

  // bootstrap / resume the widget
  // TODO: follow widget implementation to handle redirect bootstrap scenarios
  useEffect(() => {
    if (authClient.idx.canProceed()) {
      resume();
    } else {
      bootstrap();
    }
  }, [authClient, setIdxTransaction, bootstrap, resume]);

  // Derived value from idxTransaction
  const formBag = useMemo<Undefinable<FormBag>>(() => {
    // cancelled transactions will be bootstrapped again, so we wait if that happens
    if (idxTransaction === undefined || idxTransaction?.status === IdxStatus.CANCELED) {
      return undefined;
    }

    // null indicates uncaught error
    if (idxTransaction === null) {
      return transformUnhandledErrors(authApiError);
    }

    if (TERMINAL_STATUSES.includes(idxTransaction.status) || !idxTransaction.nextStep) {
      return transformTerminalTransaction(idxTransaction, widgetProps);
    }

    return transformTransaction(idxTransaction as IdxTransactionWithNextStep, widgetProps);
  // eslint-disable-next-line react-hooks/exhaustive-deps
  }, [idxTransaction]);

  // For fixing screen
  const delay = useCallback(debounce((event) => {
    setFormErrors(event.errors || []);
    setData(event.data);
  }, 100), [setFormErrors, setData]);

  const onChange = (event: Pick<JsonFormsCore, 'data' | 'errors'>) => {
    delay(event);
  };

  const handleSubmit = useCallback(async (e: any) => {
    e.preventDefault();

    // discard submission action when form is not available
    if (!formBag || !idxTransaction) {
      return;
    }

    // TODO: disable action buttons after click

    const { nextStep: { action } = {} } = idxTransaction;
    const payload = buildPayload(formBag, data);
    // Per SDK team action is guranteed to exist here, so we are safe to assert non-null
    // eslint-disable-next-line @typescript-eslint/no-non-null-assertion
    const transaction = await action!({
      ...payload,
      stateHandle: idxTransaction?.context.stateHandle,
    })
      .catch((error) => {
        setAuthApiError(error);
        console.error(error);
        return null;
      });
    setIdxTransaction(transaction);
  }, [data, formBag, idxTransaction, setIdxTransaction]);

  // Update idxTransaction when new status comes back from polling
  useEffect(() => {
    if (!idxTransaction || !pollingTransaction) {
      return;
    }

    if (pollingTransaction.nextStep?.name !== idxTransaction.nextStep?.name) {
      setIdxTransaction(pollingTransaction);
    }
  // eslint-disable-next-line react-hooks/exhaustive-deps
  }, [
    pollingTransaction, // only watch on pollingTransaction changes
  ]);

  useEffect(() => {
    if (!idxTransaction) {
      return;
    }

    // reset form data when a new transaction is available
    setData({});

    const { messages: newMessages, status } = idxTransaction;
    switch (status) {
      case (IdxStatus.PENDING): {
        // Handle global errors or warnings
        if (newMessages) {
          setMessages(newMessages);
        }
        break;
      }
      case (IdxStatus.SUCCESS):
      case (IdxStatus.TERMINAL): {
        break;
      }
      case (IdxStatus.CANCELED): {
        // clear idxTransaction to start loading state
        setIdxTransaction(undefined);
        bootstrap();
        break;
      }
      default: {
        if (newMessages) {
          // Set error.
          setMessages(newMessages);
        }
      }
    }
  // eslint-disable-next-line react-hooks/exhaustive-deps
  }, [idxTransaction, setMessages, bootstrap]);

  return (
    <WidgetContextProvider value={{
      authClient,
      data,
      formBag,
      onSuccessCallback: onSuccess,
      setIdxTransaction,
      setMessages,
      transaction: idxTransaction,
    }}
    >
      <ThemeProvider theme={mapThemeFromBrand(brandColors)}>
        <AuthContainer>
          <AuthHeader
            logo={logo}
            logoText={logoText}
            brandName={brandName}
            authCoinProps={buildAuthCoinProps(idxTransaction)}
          />
          <AuthContent>
            <InfoSection messages={messages} />
            {
              formBag ? (
                // TODO: add form onSubmit handler and assign buttons proper types
                <form
                  noValidate
                  onSubmit={handleSubmit}
                >
                  <JsonForms
                    schema={formBag.schema}
                    uischema={formBag.uischema}
                    data={data}
                    renderers={renderers}
                    cells={vanillaCells}
                    ajv={ajv}
                    onChange={onChange}
                  />
                </form>
              ) : <Spinner />
            }
          </AuthContent>
        </AuthContainer>
      </ThemeProvider>
    </WidgetContextProvider>
  );
};
