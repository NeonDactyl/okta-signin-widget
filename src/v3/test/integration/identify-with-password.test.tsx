/*
 * Copyright (c) 2022-present, Okta, Inc. and/or its affiliates. All rights reserved.
 * The Okta software accompanied by this notice is provided pursuant to the Apache License, Version 2.0 (the "License.")
 *
 * You may obtain a copy of the License at http://www.apache.org/licenses/LICENSE-2.0.
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
 * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 *
 * See the License for the specific language governing permissions and limitations under the License.
 */

import '@testing-library/jest-dom';
import {
  fireEvent,
  render,
  screen,
  waitFor,
} from '@testing-library/preact';
import { h } from 'preact';
import { OktaAuth } from '@okta/okta-auth-js';
import { Widget } from '../../src/components/Widget';
import { updateDynamicAttribute } from './util';

// import identifyWithPassword from '../../mocks/data/idp/idx/identify-with-password.json';
// This only works with new jsonTransformer
describe.skip('identify-with-password', () => {
  let authClient: OktaAuth;

  beforeAll(() => {
    // authClient = createAuthClient(identifyWithPassword);
  });

  it('renders form', async () => {
    const { container } = render(<Widget authClient={authClient} />);
    await screen.findByLabelText(/Username/);

    // NOTE: should some of these attributes be static from odyssey side?
    updateDynamicAttribute(container, ['id', 'for', 'aria-describedby', 'aria-labelledby']);

    expect(container).toMatchSnapshot();
  });

  // TODO: revisit to enable submission payload testing
  // Issue: with fireEvent.input, the event value can reach to component level onChange handler
  // but data is lost after jsonforms processing
  it.skip('sends correct payload', async () => {
    jest.spyOn(authClient.idx, 'proceed');
    render(<Widget authClient={authClient} />);

    await screen.findByLabelText(/Username/);

    const usernameEl = await screen.findByTestId('#/properties/identifier');
    const passwordEl = await screen.findByTestId('#/properties/credentials.passcode');
    const submitButton = await screen.findByTestId('identify.submit');

    fireEvent.input(usernameEl, { target: { value: 'testuser@okta.com' } });
    fireEvent.input(passwordEl, { target: { value: 'password' } });
    fireEvent.click(submitButton);

    await waitFor(() => {
      expect(authClient.options.httpRequestClient).toHaveBeenCalledTimes(3);
      expect(authClient.options.httpRequestClient).toHaveBeenNthCalledWith(3,
        'POST',
        'http://localhost:3000/idp/idx/identify',
        {
          // TODO: dfs mock response to update stateHandle field before serve
          data: '{"identifier":"testuser@okta.com","credentials":{"passcode":"password"},"stateHandle":"fake-stateHandle"}',
          headers: {
            Accept: 'application/vnd.okta.v1+json',
            'Content-Type': 'application/json',
            'X-Okta-User-Agent-Extended': 'okta-auth-js/6.5.1',
          },

          withCredentials: true,
        });
    });
  });
});
