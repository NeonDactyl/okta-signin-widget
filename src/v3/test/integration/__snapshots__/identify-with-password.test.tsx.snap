// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`identify-with-password renders form 1`] = `
<div>
  <main
    class="ods-25UsME ods-2bre13 ods-3ss0Qq ods-3TwtWX ods-4TNAee ods-7kySDp ods-155Dqk ods-58oYDe ods-1LxNEe auth-container main-container mainViewContainer"
  >
    <div
      class="ods-25UsME ods-5FzqwZ ods-3rpMjU ods-5phuZa ods-2Tkrnl ods-6DSSwZ ods-1JDDx8 ods-3DJwCa ods-4TNAee ods-7kySDp ods-155Dqk ods-58oYDe ods-1LxNEe siwContainer"
    >
      <div
        class="okta-sign-in-header auth-header siwHeader"
      >
        <h1
          class="ods-25UsME ods-155Dqk ods-3ak841"
        >
          <img
            alt="Logo"
            class="auth-org-logo siwOrgLogo"
            src="/img/atko_brand_logo.svg"
          />
        </h1>
      </div>
      <div
        class="ods-25UsME ods-2bre13 ods-5FzqwZ ods-3hyVmm ods-69Ni9p ods-5phuZa ods-2Tkrnl ods-6q66ih ods-4TNAee ods-7kySDp ods-155Dqk ods-58oYDe ods-1LxNEe"
      >
        <form
          novalidate=""
        >
          <div
            class="vertical-layout"
          >
            <div
              class="vertical-layout-item"
            >
              <div
                class="ods-25UsME ods-5Ip9Cc ods-4TNAee ods-7kySDp ods-155Dqk ods-58oYDe ods-1LxNEe"
              >
                <div
                  class="ods-25UsME ods-2bre13 ods-5FzqwZ ods-4TNAee ods-7kySDp ods-155Dqk ods-58oYDe ods-1LxNEe ods-4zKnH2"
                >
                  <label
                    class="ods-25UsME ods-2bre13 ods-4tyjL3 ods-4fFeHD ods-4TNAee ods-7kySDp ods-155Dqk ods-58oYDe ods-1LxNEe ods-g8inGu"
                    for="generated"
                  >
                    <span
                      class="ods-25UsME ods-21ITbf ods-2rdluJ ods-3j6hVs ods-ijL2fj ods-YSqZm9 ods-5vI4ZI ods-2q394D ods-T51Sls"
                    >
                      Username
                    </span>
                  </label>
                  <input
                    aria-describedby="generated"
                    class="ods-5NZW2e"
                    data-testid="#/properties/identifier"
                    id="generated"
                    name="identifier"
                    type="text"
                  />
                </div>
              </div>
            </div>
            <div
              class="vertical-layout-item"
            >
              <div
                class="ods-25UsME ods-5Ip9Cc ods-4TNAee ods-7kySDp ods-155Dqk ods-58oYDe ods-1LxNEe"
              >
                <div
                  class="ods-25UsME ods-2bre13 ods-5FzqwZ ods-4TNAee ods-7kySDp ods-155Dqk ods-58oYDe ods-1LxNEe ods-4zKnH2"
                >
                  <label
                    class="ods-25UsME ods-2bre13 ods-4tyjL3 ods-4fFeHD ods-4TNAee ods-7kySDp ods-155Dqk ods-58oYDe ods-1LxNEe ods-g8inGu"
                    for="generated"
                  >
                    <span
                      class="ods-25UsME ods-21ITbf ods-2rdluJ ods-3j6hVs ods-ijL2fj ods-YSqZm9 ods-5vI4ZI ods-2q394D ods-T51Sls"
                    >
                      Password
                    </span>
                  </label>
                  <input
                    aria-describedby="generated"
                    class="ods-5NZW2e"
                    data-testid="#/properties/credentials.passcode"
                    id="generated"
                    name="credentials.passcode"
                    type="password"
                  />
                </div>
                <button
                  class="ods-25UsME ods-155Dqk ods-EdLvMR ods-22KTj5"
                  type="button"
                >
                  <span
                    class="ods-2V59HN"
                  >
                    renderers.password.showPassword
                  </span>
                </button>
              </div>
            </div>
            <div
              class="vertical-layout-item"
            >
              <div
                class="ods-25UsME ods-5Ip9Cc ods-4TNAee ods-7kySDp ods-155Dqk ods-58oYDe ods-1LxNEe"
              >
                <div
                  class="ods-25UsME ods-2L0e0z ods-4TNAee ods-7kySDp ods-155Dqk ods-58oYDe ods-1LxNEe"
                >
                  <input
                    aria-describedby="generated"
                    class="ods-4fV8d7"
                    id="generated"
                    type="checkbox"
                  />
                  <label
                    class="ods-1gChEk"
                    for="generated"
                  >
                    <span
                      class="ods-IDodlr"
                      role="presentation"
                    >
                      <span
                        class="ods-6H0I0R"
                      >
                        <svg
                          aria-labelledby="generated"
                          class="ods-2icygl"
                          fill="none"
                          role="presentation"
                          viewBox="0 0 14 14"
                          xmlns="http://www.w3.org/2000/svg"
                        >
                          <path
                            d="M0 6.53333L1.86667 4.66667L5.6 8.51667L12.1333 1.86667L14 3.61667L5.6 12.25L0 6.53333Z"
                            fill="currentColor"
                          />
                        </svg>
                      </span>
                    </span>
                    Keep me signed in
                  </label>
                </div>
              </div>
            </div>
            <div
              class="vertical-layout-item"
            >
              <div
                class="ods-25UsME ods-30mmIB ods-4TNAee ods-7kySDp ods-155Dqk ods-58oYDe ods-1LxNEe"
              >
                <button
                  class="ods-25UsME ods-155Dqk ods-EdLvMR ods-2VYSC6"
                  data-testid="identify.submit"
                  type="submit"
                >
                  <span
                    class="ods-2V59HN"
                  >
                    Submit
                  </span>
                </button>
              </div>
            </div>
            <div
              class="vertical-layout-item"
            >
              <div
                class="ods-25UsME ods-30mmIB ods-4TNAee ods-7kySDp ods-155Dqk ods-58oYDe ods-1LxNEe"
              >
                <button
                  class="ods-25UsME ods-155Dqk ods-EdLvMR ods-2VYSC6"
                  data-testid="select-enroll-profile.cta"
                  type="button"
                >
                  <span
                    class="ods-2V59HN"
                  >
                    select-enroll-profile
                  </span>
                </button>
              </div>
            </div>
            <div
              class="vertical-layout-item"
            >
              <div
                class="ods-25UsME ods-30mmIB ods-4TNAee ods-7kySDp ods-155Dqk ods-58oYDe ods-1LxNEe"
              >
                <button
                  class="ods-25UsME ods-155Dqk ods-EdLvMR ods-2VYSC6"
                  data-testid="currentAuthenticator-recover.cta"
                  type="button"
                >
                  <span
                    class="ods-2V59HN"
                  >
                    currentAuthenticator-recover
                  </span>
                </button>
              </div>
            </div>
            <div
              class="vertical-layout-item"
            >
              <div
                class="ods-25UsME ods-30mmIB ods-4TNAee ods-7kySDp ods-155Dqk ods-58oYDe ods-1LxNEe"
              >
                <button
                  class="ods-25UsME ods-155Dqk ods-EdLvMR ods-2VYSC6"
                  data-testid="cancel.cta"
                  type="button"
                >
                  <span
                    class="ods-2V59HN"
                  >
                    cancel
                  </span>
                </button>
              </div>
            </div>
          </div>
        </form>
      </div>
    </div>
  </main>
</div>
`;
